app/src/server/tasks/generateInvoices.task.ts

```diff
@@ -0,0 +1,180 @@
+      .insertInto("Invoice")
+      .values({
+        id: uuidv4(),
+        projectId: projectId,
+        amount: 0,
+        slug: getInvoiceSlug(),
+        billingPeriod: getPreviousMonthWithYearString(),
+      })
+      .returning(["id", "createdAt", "slug"])
+      .executeTakeFirst();
+
+    // 2. Associate stats with invoice
+    if (invoice) {
+      await tx
+        .updateTable("UsageLog")
+        .set({
+          invoiceId: invoice.id,
+        })
+        .where("projectId", "=", projectId)
+        .where("createdAt", ">=", startDate)
+        .where("createdAt", "<=", endDate)
+        .execute();
+
+      // 3. Calculate usage stats
+      const stats = await getStats(
+        tx.selectFrom("UsageLog as ul").where("ul.invoiceId", "=", invoice.id),
+      );
+
+      // 4. Calculate available credits
+      const creditsAvailable = await tx
+        .selectFrom("CreditAdjustment as ca")
+        .where("ca.projectId", "=", projectId)
+        .where("ca.createdAt", "<=", endDate)
+        .select(({ fn }) => fn.sum(sql<number>`amount`).as("amount"))
+        .executeTakeFirst();
+
+      if (stats && creditsAvailable) {
+        // 5. Calculate total spend
+        const { totalSpent, creditsUsed, remainingCredits } = calculateSpendingsWithCredits(
+          Number(stats?.cost ?? 0),
+          Number(creditsAvailable.amount ?? 0),
+        );
+
+        // 5. Update invoice
+        await tx
+          .updateTable("Invoice")
+          .set({
+            amount: totalSpent,
+            status: totalSpent >= 1 ? "PENDING" : "CANCELLED", // Minimum $1 charge
+            description: JSON.stringify(
+              getInvoiceDescription({
+                totalInferenceSpend: Number(stats?.totalInferenceSpend ?? 0),
+                totalTrainingSpend: Number(stats?.totalTrainingSpend ?? 0),
+                totalInputTokens: Number(stats?.totalInputTokens ?? 0),
+                totalOutputTokens: Number(stats?.totalOutputTokens ?? 0),
+                totalTrainingTokens: Number(stats?.totalTrainingTokens ?? 0),
+                creditsUsed,
+                remainingCredits,
+              }),
+            ),
+          })
+          .where("id", "=", invoice.id)
+          .execute();
+
+        // 6. Create negative credit adjustment if credits were used
+        if (creditsUsed > 0) {
+          await tx
+            .insertInto("CreditAdjustment")
+            .values({
+              id: uuidv4(),
+              projectId: projectId,
+              amount: -creditsUsed,
+              //Adjustments should be created on the first day of the next month, when invoices are generated
+              createdAt: dayjs(endDate).add(1, "month").startOf("month").toDate(),
+              description: `Invoice #${invoice.slug}`,
+              invoiceId: invoice.id,
+              type: "INVOICE",
+            })
+            .execute();
+        }
+      }
+    }
+  });
+}
+
+function getInvoiceSlug(length = 6) {
```

Comment:
Nit: I would use "generate" instead of "get" in the names of these functions. "get" implies that we're retrieving information that already exists, whereas "generate" implies that we're creating a new value.

Reply:
I see, nice idea!